### ISSUE TITLE
Strange behavior of AbsentLink

### ISSUE THREAD

**Issue by shujingke** [opencog/atomspace#1596]
AbsentLink evaluates if a term has any matches in the Atomspace. According to this page: 
<https://wiki.opencog.org/w/Atom_relationships> ,  AbsentLink  should work
for both patterns with variables or constant terms. But the output of running the "test-absent-rule" below looks not normal:

```
 ;; Knowledge base:
(Inheritance (stv 1.0 1.0)
   (Concept "American")
   (Concept "person"))

(Inheritance (stv 1.0 1.0)
   (Concept "German")
   (Concept "person"))

(Inheritance (stv 1.0 1.0)
   (Concept "cat")
   (Concept "pet"))

(Inheritance (stv 1.0 1.0)
   (Concept "dog")
   (Concept "pet"))


(Evaluation (stv 1.0 1.0)
   (Predicate "keep-pet")
   (List
   	(Concept "German")
        (Concept "dog")))

(define test-absent-rule
 (let* (
          (kp (Predicate "keep-pet"))
          (vA (Variable "$A"))        
          (vX (Variable "$X")) 
             
       )

  (BindLink
   (VariableList
     (TypedVariable
        vA
        (Type "ConceptNode"))
     (TypedVariable
        vX
        (Type "ConceptNode"))
  
   )
   (And
     (Inheritance
        vA
        (Concept "person"))
     (Inheritance
        vX
        (Concept "pet"))


     ;; test if it already exist
     (Absent 
        (Evaluation kp (List vA vX))
     )
 
   )     
 
   (cog-set-tv! (Evaluation kp (List vA vX)) (stv 1 1))
  
  )
 )
)

// run  cog-execute! test-absent-rule, output:
(SetLink
    (AndLink
       (EvaluationLink (stv 1 1)
          (PredicateNode "keep-pet")
          (ListLink
             (ConceptNode "American")
             (VariableNode "$X")
          )
       )
    )
)

```

**Comment by linas** [comment]
Oof. There are several issues here. 

1) first, you understand that the `(cog-set-tv!` will run when its typed in, and not when `cog-execute!` runs, right?  It is just going to set the TV on the evaluation link with variables in it.

2) As designed, that BindLink creates exactly the thing that it is testing for absence, so if you run it a second time, it will no longer be absent.  Is that really what you want to do?

3) Yes, there is a bug somewhere, that output that you show is wrong.  I think I know why, and will try to fix it.

**Comment by linas** [comment]
Hmm Fixing this correctly requires treaing up a lot of (old) code, I won't do this today.  Basically, the code for AbsentLink is 8 years old, dating back to the early days of the pattern matcher. Its based on a flawed design; the correct solution is to treat it (more or less) like any other virtual link, but this requires a fairly large redesign, because it touches a lot of places.  I cannot do it in a few hours.

**Comment by shujingke** [comment]
I am using something like :
`(bool->tv (null? (cog-link "EvaluationLink" predicate (List A B))) )`
to check existence now and it works.

From the other issue (#1598), you said PresentLink can be used for checking existence, which is nice, and how to use only PresentLink (not AbsentLink) to check non-existence? For example, in the above example:
```
     (Absent 
        (Evaluation kp (List vA vX))
     )
```
Can it be achieved by:
```
     (Not 
         (PresentLink 
              (Evaluation kp (List vA vX))
         )
     )
```

**Comment by linas** [comment]
As I mentioned: the AbsentLink is very old, and uses very old code paths. The PresentLink is new, it was added maybe 2-3 years ago, and it incompletely  supported. It is good that `(Not (PresentLink))` works, but that is only because you got lucky: the implementation is incomplete, and there are likely to be more bugs.  The simplest cases work. The more complex ones do not.  

In fact, for this example, I believe that `(Not (PresentLink))` is returning the right answer only by accident, since I am sure that is not (cannot be) taking the correct code path.

**Comment by linas** [comment]
Re `(null? (cog-link "EvaluationLink" predicate (List A B)))`  Yes! this is exactly what `coglink` was designed to do: to tell you simply, yes or no, whether some atom exists, or does not exist, in the atomspace (without accidentally creating that atom).

The AbsentLink was designed for more complex queries that cannot be reduced to just single atoms.

**Comment by linas** [comment]
FYI, I've been working on this on and off for the last week, and have a partial solution, but its taking a fair bit more time than I expected.

**Comment by linas** [comment]
Status update: I've written a lot of code for this; its in a private branch; (abs, in my repo) its unexpectedly complicated.  I have vague plans to return to this, if I cannot convince @vsbogd to work on it first.

**Comment by linas** [comment]
Half-finished fix is here: https://github.com/linas/atomspace/tree/absent-link-rework -- fixing this was not as easy as it first appears.

**Comment by linas** [comment]
Closing, This is fixed by pull req #2174 -- which makes `Not Present` behave the same as `Absent`

**Comment by linas** [comment]
Umm, no there is still some bug.

**Comment by linas** [comment]
Closing. This is fixed by #2179
